tests/components/tasmota/test_binary_sensor.py

```diff
@@ -0,0 +1,222 @@
+    assert not state.attributes.get(ATTR_ASSUMED_STATE)
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/LWT", "Online")
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_OFF
+    assert not state.attributes.get(ATTR_ASSUMED_STATE)
+
+    # Test normal state update
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1", '{"STATE":"ON"}')
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_ON
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1", '{"STATE":"OFF"}')
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_OFF
+
+    # Test periodic state update
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/SENSOR", '{"Switch1":"ON"}')
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_ON
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/SENSOR", '{"Switch1":"OFF"}')
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_OFF
+
+    # Test polled state update
+    async_fire_mqtt_message(
+        hass, "tasmota_49A3BC/stat/STATUS8", '{"StatusSNS":{"Switch1":"ON"}}'
+    )
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_ON
+
+    async_fire_mqtt_message(
+        hass, "tasmota_49A3BC/stat/STATUS8", '{"StatusSNS":{"Switch1":"OFF"}}'
+    )
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_OFF
+
+
+async def test_off_delay(hass, mqtt_mock, setup_tasmota):
+    """Test off_delay option."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    config["swc"][0] = 13  # PUSHON: 1s off_delay
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    events = []
+
+    @ha.callback
+    def callback(event):
+        """Verify event got called."""
+        events.append(event.data["new_state"].state)
+
+    hass.bus.async_listen(EVENT_STATE_CHANGED, callback)
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/tele/LWT", "Online")
+    await hass.async_block_till_done()
+    assert events == ["off"]
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1", '{"STATE":"ON"}')
+    await hass.async_block_till_done()
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_ON
+    assert events == ["off", "on"]
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1", '{"STATE":"ON"}')
+    await hass.async_block_till_done()
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_ON
+    assert events == ["off", "on", "on"]
+
+    async_fire_time_changed(hass, dt_util.utcnow() + timedelta(seconds=1))
+    await hass.async_block_till_done()
+    state = hass.states.get("binary_sensor.test")
+    assert state.state == STATE_OFF
+    assert events == ["off", "on", "on", "off"]
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6299130)